src/main/java/com/conveyal/r5/streets/TarjanIslandPruner.java

```diff
@@ -0,0 +1,326 @@
+                            }
+                        });
+                    }
+                }
+            }
+        }
+
+        LOG.info("Found {} strong components for mode {} using Tarjan's algorithm in {}sec",
+                strongComponents.size(), mode, (System.currentTimeMillis() - startTime) / 1000d);
+
+        startTime = System.currentTimeMillis();
+
+        int nComponentsRemoved = 0;
+        int nVerticesRemoved = 0;
+
+        for (TIntSet strongComponent : strongComponents) {
+            if (strongComponent.size() < minComponentSize) {
+                nComponentsRemoved++;
+                for (TIntIterator it = strongComponent.iterator(); it.hasNext();) {
+                    int vertex = it.next();
+                    nVerticesRemoved++;
+                    removePermissionsAroundVertex(vertex);
+                }
+            }
+        }
+
+        LOG.info("Removed {} strong component (islands) with fewer than {} vertices for mode {} in {}sec. {} vertices removed.",
+                nComponentsRemoved, minComponentSize, mode, (System.currentTimeMillis() - startTime) / 1000d, nVerticesRemoved);
+    }
+
+    /** Loop over every outgoing edge for a particular mode */
+    public void forEachOutgoingEdge (int vertex, Consumer<EdgeStore.Edge> consumer) {
+        streets.outgoingEdges.get(vertex).forEach(eidx -> {
+            edgeCursor.seek(eidx);
+
+            // filter by mode
+            switch (mode) {
+                case WALK:
+                    if (!edgeCursor.getFlag(EdgeStore.EdgeFlag.ALLOWS_PEDESTRIAN)) return true;
+                    else break;
+                case CAR:
+                    if (!edgeCursor.getFlag(EdgeStore.EdgeFlag.ALLOWS_CAR)) return true;
+                    else break;
+                case BICYCLE:
+                    // include ped mode here, because walking bikes is a thing you can do.
+                    boolean allowsBike = edgeCursor.getFlag(EdgeStore.EdgeFlag.ALLOWS_BIKE) ||
+                                    edgeCursor.getFlag(EdgeStore.EdgeFlag.ALLOWS_PEDESTRIAN);
+                    if (!allowsBike) return true;
+                    else break;
+                default:
+                    throw new IllegalArgumentException(String.format("Unsupported mode %s for island removal", mode));
+            }
+
+            consumer.accept(edgeCursor);
+            return true; // continue iteration over outgoing edges
+        });
+    }
+
+    /** Remove the permissions around a vertex for the desired mode. Returns the number of edges affected */
+    public void removePermissionsAroundVertex (int vertex) {
+        for (TIntList edgeList : new TIntList[] { streets.outgoingEdges.get(vertex), streets.incomingEdges.get(vertex) }) {
+            edgeList.forEach(eidx -> {
+                edgeCursor.seek(eidx);
+                switch (mode) {
```

Comment:
As mentioned above, this switch could probably be pulled out into a method on Edge to make things more readable.